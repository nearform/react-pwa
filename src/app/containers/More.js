const React = require('react')
const Link = require('../components/Link')
const { withRouter } = require('react-router')
const { px } = require('csx')
const { style } = require('typestyle')
const { debugClassName } = require('../styles/common')

const buildLink = pathname => {
  let pathParts = pathname.split('/')
  let currentPage = parseInt(pathname.split('page/')[1])
  let nextPage = Number.isInteger(currentPage) ? currentPage + 1 : 2
  if (!pathParts[1] || pathParts[1] === 'page') return `/page/${nextPage}`
  return `/${pathParts[1]}/page/${nextPage}`
}

const moreLinkClassName = style(
  debugClassName('more-link'),
  {
    marginBottom: px(6),
    padding: `0 0 0 ${px(30)}`
  }
)

const More = props => {
  let link = buildLink(props.location.pathname)
  return (
    <p className={moreLinkClassName}>
      <Link to={link}>More</Link>
    </p>
  )
}

module.exports = withRouter(More)
