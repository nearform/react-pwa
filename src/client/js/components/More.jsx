import React from 'react'
import { Link } from 'react-router-dom'
import { stylesheet, classes } from 'typestyle'
import { colors } from '../styles/common'

const buildLinks = pathname => {
  const pathParts = pathname.split('/')
  const currentPage = parseInt(pathname.split('page/')[1])
  const nextPage = Number.isInteger(currentPage) ? currentPage + 1 : 2
  const prevPage = Number.isInteger(currentPage) && currentPage > 2 ? currentPage - 1 : null

  if (!pathParts[1] || pathParts[1] === 'page') {
    return {
      currentPage,
      prevLinkEnabled: currentPage >= 2,
      prevLink: Number.isInteger(currentPage) && currentPage > 2 ? `/page/${prevPage}` : '/',
      nextLink: `/page/${nextPage}`
    }
  }

  return {
    currentPage,
    prevLinkEnabled: currentPage >= 2,
    prevLink: Number.isInteger(currentPage) && currentPage > 2 ? `/${pathParts[1]}/page/${prevPage}` : `/${pathParts[1]}`,
    nextLink: `/${pathParts[1]}/page/${nextPage}`
  }
}

const styles = stylesheet({
  more: {
    padding: `1em 0 1em 0`,
    background: colors.LIGHTEST_GRAY,
    textAlign: 'right',
    gridColumnStart: '1',
    gridColumnEnd: '4',
    gridRowStart: '2',
    marginRight: '6px'
  },
  moreItem: {
    background: colors.NEARFORM_BRAND_ACCENT_2,
    marginLeft: '1em',
    color: 'white',
    cursor: 'pointer',
    fontWeight: '800',
    padding: '10px 12px',
    borderRadius: '10px',
    boxShadow: '2px 2px 8px 0px rgba( 0, 0, 0, 0.2 )'
  },
  moreDisabled: {
    background: colors.GRAY,
    cursor: 'not-allowed',
    boxShadow: 'none'
  }
})

export function More({ location }) {
  const {nextLink, prevLink, prevLinkEnabled, currentPage} = buildLinks(location.pathname)

  const rangeStart = currentPage ? 30 * (currentPage - 1) + 1 : 1
  const rangeEnd = rangeStart + 29

  return (
    <div className={styles.more}>
      <span>{rangeStart} - {rangeEnd}</span>

      {prevLinkEnabled &&
        <Link className={styles.moreItem} to={prevLink}>
          {'<'}
        </Link>
      }

      {!prevLinkEnabled &&
        <span className={classes(styles.moreItem, styles.moreDisabled)}>
          {'<'}
        </span>

      }

      <Link className={styles.moreItem} to={nextLink}>
        {'>'}
      </Link>
    </div>
  )
}
