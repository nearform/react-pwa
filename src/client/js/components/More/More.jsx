import React from 'react'
import { Link } from 'react-router-dom'
import { stylesheet, classes } from 'typestyle'
import { colors } from '../../styles/common'

const PAGE_SIZE = 30 // TODO: this should be shared config and either passed in request or shared with server

const buildLinks = (pathname, currentCount) => {
  const pathParts = pathname.split('/')
  const currentPage = parseInt(pathname.split('page/')[1]) || 1
  const nextPage = Number.isInteger(currentPage) ? currentPage + 1 : 2
  const prevPage = Number.isInteger(currentPage) && currentPage > 2 ? currentPage - 1 : null

  if (!pathParts[1] || pathParts[1] === 'page') {
    return {
      currentPage,
      prevLinkEnabled: currentPage >= 2,
      nextLinkEnabled: !(currentCount < PAGE_SIZE),
      prevLink: Number.isInteger(currentPage) && currentPage > 2 ? `/page/${prevPage}` : '/',
      nextLink: `/page/${nextPage}`,
    }
  }

  return {
    currentPage,
    prevLinkEnabled: currentPage >= 2,
    nextLinkEnabled: !(currentCount < PAGE_SIZE),
    prevLink:
      Number.isInteger(currentPage) && currentPage > 2 ? `/${pathParts[1]}/page/${prevPage}` : `/${pathParts[1]}`,
    nextLink: `/${pathParts[1]}/page/${nextPage}`,
  }
}

const styles = stylesheet({
  more: {
    background: colors.LIGHTEST_GRAY,
    textAlign: 'right',
    gridColumnStart: '1',
    gridColumnEnd: '4',
    gridRowStart: '2',
    padding: '1em',
    borderBottom: `6px solid ${colors.NEARFORM_BRAND_MAIN}`,
    position: 'sticky',
    zIndex: 1,
    top: '70px', // Height of the header TODO: should be shared constant
  },
  moreItem: {
    background: colors.NEARFORM_BRAND_ACCENT_2,
    marginLeft: '1em',
    color: 'white',
    cursor: 'pointer',
    fontWeight: '800',
    padding: '10px 12px',
    borderRadius: '10px',
    boxShadow: '2px 2px 8px 0px rgba( 0, 0, 0, 0.2 )',
  },
  moreDisabled: {
    background: colors.GRAY,
    cursor: 'not-allowed',
    boxShadow: 'none',
  },
})

export function More({ location, data }) {
  const currentCount = data ? data.length : 0
  const { nextLink, prevLink, nextLinkEnabled, prevLinkEnabled, currentPage } = buildLinks(
    location.pathname,
    currentCount
  )

  const rangeStart = currentPage ? 30 * (currentPage - 1) + 1 : 1
  const rangeEnd = rangeStart + 29

  return (
    <div className={styles.more}>
      <span>{`${rangeStart} - ${rangeEnd}`}</span>

      <Link
        className={classes(styles.moreItem, !prevLinkEnabled && styles.moreDisabled)}
        role="button"
        aria-label="Load previous posts"
        disabled={!prevLinkEnabled}
        to={prevLink}
      >
        {'<'}
      </Link>

      <Link
        className={classes(styles.moreItem, !nextLinkEnabled && styles.moreDisabled)}
        role="button"
        aria-label="Load next posts"
        disabled={!nextLinkEnabled}
        to={nextLink}
      >
        {'>'}
      </Link>
    </div>
  )
}
