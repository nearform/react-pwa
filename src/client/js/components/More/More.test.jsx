import React from 'react'
import { render } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'

import { More } from './More'

it('renders correctly when location is "/"', () => {
  const props = {
    location: {
      pathname: '/',
    },
    data: Array(30),
  }

  const test = render(
    <MemoryRouter initialEntries={[props.location.pathname]}>
      <More {...props} />
    </MemoryRouter>
  )

  expect(test).toMatchSnapshot()
})

it('renders correctly when location is "/page/5"', () => {
  const props = {
    location: {
      pathname: '/page/5',
    },
    data: Array(30),
  }

  const test = render(
    <MemoryRouter initialEntries={[props.location.pathname]}>
      <More {...props} />
    </MemoryRouter>
  )

  expect(test).toMatchSnapshot()
})

it('renders correctly when location is "/newcomments"', () => {
  const props = {
    location: {
      pathname: '/newcomments',
    },
    data: Array(30),
  }

  const test = render(
    <MemoryRouter initialEntries={[props.location.pathname]}>
      <More {...props} />
    </MemoryRouter>
  )

  expect(test).toMatchSnapshot()
})

it('renders correctly when location is "/newcomments/page/2"', () => {
  const props = {
    location: {
      pathname: '/newcomments/page/2',
    },
    data: Array(30),
  }

  const test = render(
    <MemoryRouter initialEntries={[props.location.pathname]}>
      <More {...props} />
    </MemoryRouter>
  )

  expect(test).toMatchSnapshot()
})

it('renders correctly when end of list reached', () => {
  const props = {
    location: {
      pathname: '/newcomments/page/2',
    },
    data: Array(29),
  }

  const test = render(
    <MemoryRouter initialEntries={[props.location.pathname]}>
      <More {...props} />
    </MemoryRouter>
  )

  expect(test).toMatchSnapshot()
})
