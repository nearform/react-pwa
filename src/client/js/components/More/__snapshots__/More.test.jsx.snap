// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when end of list reached 1`] = `
<div
  class="more_f1wasfrc"
>
  <span>
    31 - 60
  </span>
  <a
    class="moreItem_f8pzcup"
    href="/newcomments"
  >
    &lt;
  </a>
  <span
    class="moreItem_f8pzcup moreDisabled_fs9on40"
  >
    &gt;
  </span>
</div>
`;

exports[`renders correctly when location is "/" 1`] = `
<div
  class="more_f1wasfrc"
>
  <span>
    1 - 30
  </span>
  <span
    class="moreItem_f8pzcup moreDisabled_fs9on40"
  >
    &lt;
  </span>
  <a
    class="moreItem_f8pzcup"
    href="/page/2"
  >
    &gt;
  </a>
</div>
`;

exports[`renders correctly when location is "/newcomments" 1`] = `
<div
  class="more_f1wasfrc"
>
  <span>
    1 - 30
  </span>
  <span
    class="moreItem_f8pzcup moreDisabled_fs9on40"
  >
    &lt;
  </span>
  <a
    class="moreItem_f8pzcup"
    href="/newcomments/page/2"
  >
    &gt;
  </a>
</div>
`;

exports[`renders correctly when location is "/newcomments/page/2" 1`] = `
<div
  class="more_f1wasfrc"
>
  <span>
    31 - 60
  </span>
  <a
    class="moreItem_f8pzcup"
    href="/newcomments"
  >
    &lt;
  </a>
  <a
    class="moreItem_f8pzcup"
    href="/newcomments/page/3"
  >
    &gt;
  </a>
</div>
`;

exports[`renders correctly when location is "/page/5" 1`] = `
<div
  class="more_f1wasfrc"
>
  <span>
    121 - 150
  </span>
  <a
    class="moreItem_f8pzcup"
    href="/page/4"
  >
    &lt;
  </a>
  <a
    class="moreItem_f8pzcup"
    href="/page/6"
  >
    &gt;
  </a>
</div>
`;
